### PR TITLE
Add AlfvenWave solution

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

An AnalyticSolution to the GrMhd system:
a circularly polarized Alfven wave travelling in the z direction.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.hpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  AlfvenWave& operator=(AlfvenWave&& /*rhs*/) noexcept = default;
+  ~AlfvenWave() = default;
+
+  AlfvenWave(WaveNumber::type wavenumber, Pressure::type pressure,
+             RestMassDensity::type rest_mass_density,
+             AdiabaticExponent::type adiabatic_exponent,
+             BackgroundMagField::type background_mag_field,
+             PerturbationSize::type perturbation_size) noexcept;
+
+  explicit AlfvenWave(CkMigrateMessage* /*unused*/) noexcept {}
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.hpp]: Is this constructor necessary? (If it compiles without it, let's leave it out)

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.hpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+      std::numeric_limits<double>::signaling_NaN();
+  BackgroundMagField::type background_mag_field_ =
+      std::numeric_limits<double>::signaling_NaN();
+  PerturbationSize::type perturbation_size_ =
+      std::numeric_limits<double>::signaling_NaN();
+  double alfven_speed_ = std::numeric_limits<double>::signaling_NaN();
+  double fluid_speed_ = std::numeric_limits<double>::signaling_NaN();
+};
+
+inline bool operator==(const AlfvenWave& lhs, const AlfvenWave& rhs) noexcept {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.hpp]: I'm inclined to move the operators to the cpp file.

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  const DataType phase = k_dot_x_minus_vt(x, t);
+  get(get<Tags::dt<hydro::Tags::SpecificInternalEnergy<DataType>>>(result)) =
+      -pressure_ /
+      ((adiabatic_exponent_ - 1.0) *
+       square(1.0 + perturbation_size_ * sin(phase))) *
+      get(get<Tags::dt<hydro::Tags::RestMassDensity<DataType>>>(result));
+  // Angular frequency:
+  const double omega = alfven_speed_ * wavenumber_;
+
+  get<Tags::dt<hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>>(
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp:105]: It looks to me like the `.get`  would fit on the previous line. Is this a clang format being silly thing?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I'm happy for you to squash

**Review by nilsdeppe — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                       const AdiabaticExponent::type adiabatic_exponent,
+                       const BackgroundMagField::type background_mag_field,
+                       const PerturbationSize::type perturbation_size) noexcept
+    : wavenumber_(wavenumber),
+      pressure_(pressure),
+      rest_mass_density_(rest_mass_density),
+      adiabatic_exponent_(adiabatic_exponent),
+      background_mag_field_(background_mag_field),
+      perturbation_size_(perturbation_size) {
+  const double rho_zero_times_h =
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp]: why define this?

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  auto result = make_with_value<
+      tuples::tagged_tuple_from_typelist<AlfvenWave::dt_variables_t<DataType>>>(
+      x, 0.0);
+
+  const DataType phase = k_dot_x_minus_vt(x, t);
+  get(get<Tags::dt<hydro::Tags::SpecificInternalEnergy<DataType>>>(result)) =
+      -pressure_ /
+      ((adiabatic_exponent_ - 1.0) *
+       square(1.0 + perturbation_size_ * sin(phase))) *
+      get(get<Tags::dt<hydro::Tags::RestMassDensity<DataType>>>(result));
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp]: I don't believe this expression can possibly be true.  Isn't it zero?  In fact you say this later.  😕

Comment (kidder) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp]: How can your test pass?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp (hunk 5 above)
Comment (mar-celine) [reply src/PointwiseFunctions/AnalyticSolutions/GrMhd/AlfvenWave.cpp]: The only non-zero quantities should be the magnetic field and the velocity field, which vary sinusoidally - this quantity ends up being zero so I shouldn't be setting it at all - thanks for catching this!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe anything for this PR?

**Review by wthrowe — APPROVED** [review]

**Pull request merged**